R/test_result.R

```diff
@@ -1,127 +1,143 @@
 #'
-#' Collect a set of test to execute against a user's result value
-#' @param ... a set of functions that will accept the evaluated user solution.
-#'   If the test fails, it should throw an error with the message to display.
-#' @noRd
-#' @rdname test
-#' @examples
+#' test_result is very similar to check_result.
+#' in check_result, we just go through all the cases and once one of the
+#' conditions match, we are done.
+#' However, in test_result, we need to go though all the conditions
+#' and store their values to tally up the total number of "good" passing cases.
+#' It's behaviour is similar to running a unit testing suite,
+#' all the cases need to be run and tallied up in the end.
 #'
-#' tests(
-#'   function(your_answer) {
-#'     checkmate::expect_function(your_answer, args = c("x"))
-#'   },
-#'   test(
-#'     # use a custom error message
-#'     "Make sure your function returns a number!",
-#'     function(your_answer) {
-#'       checkmate::expect_number(your_answer(1))
-#'     }
-#'   ),
-#'   function(your_answer) {
-#'     testthat::expect_equal(your_answer(0), NaN)
-#'   },
-#'   function(your_answer) {
-#'     testthat::expect_equal(your_answer(1:10), sqrt(log(1:10)))
-#'   }
-#' )
+#' Since the API for test_result uses pass_if and fail_if,
+#' we are presented with another problem,
+#' When a pass_if condition is found (i.e., matched and returns TRUE),
+#' that means the test is "passing".
+#' However, when a fail_if condition is found,
+#' that means the test is actually "failing",
+#' So we need to flip the `correct` condition in the graded object.
 #'
-#' \dontrun{grading_demo()}
-grader_tests <- function(...) {
-  fns <- list(...)
-  lapply(fns, function(fn) {
-    checkmate::assert_function(fn)
-    if (length(formals(fn)) == 0) {
-      stop("The function must be able to accept the user submission")
+#' This function just flips that boolean condition given depending on
+#' whether a pass_if or fail_if condition is passing or failing.
+#'
+#' This way we can store all the graded$correct values into a boolean vector.
+#' In order to calculate the number of passing conditions,
+#' we can then sum the boolean vector.
+#' @noRd
+pass_fail_condition_modify <- function(condi, grader_args, learnr_args){
+  evaluated_condi <- evaluate_condition(condi, grader_args, learnr_args)
+
+  # need to account for the case when fail_if does not match (this means the test passed)
+  # so we would need to flip the graded class correct status
+  if (condi$correct) { # evaluating a pass_if condition # nolint
+    # if a pass_if returns a NULL, it means the condition evaluated FALSE, which is a bad thing
+    # if it is NULL, we give it an "incorrect" graded class value
+    # if it is "correct", we keep the "correct" graded class value
+    evaluated_condi <- evaluated_condi %||% graded(correct = FALSE, message = NULL)
+    return(evaluated_condi)
+
+  } else { # evaluating a fail_if condition # nolint
+    if (is.null(evaluated_condi)) {
```

Comment:
I believe this if/else can be replaced with 

```r
evaluated_condi <- evaluated_condi %||% graded(correct = TRUE, message = NULL)
```

A _passing_ `fail_if` returns a `graded(correct = FALSE, message = MESSAGE)` object.

Reply:
oh you're right! I just wrote it super explicitly to not make my brain hurt.
I should update the docs about that too...